playground/backend/internal/code_processing/code_processing.go

```diff
@@ -268,56 +334,75 @@ func finishByTimeout(ctx context.Context, pipelineId uuid.UUID, cacheService cac
+	logger.Errorf("%s: Compile: err: %s, output: %s\n", pipelineId, err.Error(), errorOutput)
 
-		cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_PREPARATION_ERROR)
-	case pb.Status_STATUS_COMPILE_ERROR:
-		logger.Errorf("%s: Compile: err: %s, output: %s\n", pipelineId, err.Error(), data)
+	cacheService.SetValue(ctx, pipelineId, cache.CompileOutput, "error: "+err.Error()+", output: "+string(errorOutput))
+	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_COMPILE_ERROR)
+}
 
-		cacheService.SetValue(ctx, pipelineId, cache.CompileOutput, "error: "+err.Error()+", output: "+string(data))
+// processRunError processes error received during processing run step via setting a corresponding status and output to cache
+func processRunError(ctx context.Context, errorChannel chan error, errorOutput []byte, pipelineId uuid.UUID, cacheService cache.Cache, stopReadLogsChannel, finishReadLogsChannel chan bool) {
+	err := <-errorChannel
+	logger.Errorf("%s: Run(): err: %s, output: %s\n", pipelineId, err.Error(), errorOutput)
 
-		cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_COMPILE_ERROR)
-	case pb.Status_STATUS_RUN_ERROR:
-		logger.Errorf("%s: Run: err: %s, output: %s\n", pipelineId, err.Error(), data)
+	cacheService.SetValue(ctx, pipelineId, cache.RunError, "error: "+err.Error()+", output: "+string(errorOutput))
 
-		cacheService.SetValue(ctx, pipelineId, cache.RunError, "error: "+err.Error()+", output: "+string(data))
+	stopReadLogsChannel <- true
+	<-finishReadLogsChannel
 
-		cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_RUN_ERROR)
-	}
+	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_RUN_ERROR)
 }
 
-// processSuccess processes case after successful code processing via setting a corresponding status and output to cache
-func processSuccess(ctx context.Context, output []byte, pipelineId uuid.UUID, cacheService cache.Cache, status pb.Status) {
-	switch status {
-	case pb.Status_STATUS_PREPARING:
-		logger.Infof("%s: Validate() finish\n", pipelineId)
+// processValidateSuccess processes case after successful validate step via setting a corresponding status and output to cache
+func processValidateSuccess(ctx context.Context, pipelineId uuid.UUID, cacheService cache.Cache) {
+	logger.Infof("%s: Validate() finish\n", pipelineId)
 
-		cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_PREPARING)
-	case pb.Status_STATUS_COMPILING:
-		logger.Infof("%s: Prepare() finish\n", pipelineId)
+	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_PREPARING)
+}
+
+// processPrepareSuccess processes case after successful prepare step via setting a corresponding status and output to cache
+func processPrepareSuccess(ctx context.Context, pipelineId uuid.UUID, cacheService cache.Cache) {
+	logger.Infof("%s: Prepare() finish\n", pipelineId)
+
+	cacheService.SetValue(ctx, pipelineId, cache.Status, pb.Status_STATUS_COMPILING)
+}
```

Comment:
the same with success

Reply:
If too many arguments you can wrap them in additional structure

Reply:
Changed.